10/alpine/Dockerfile

```diff
@@ -34,6 +33,23 @@ RUN addgroup -g 1000 node \
     gpg --batch --keyserver hkp://ipv4.pool.sks-keyservers.net --recv-keys "$key" || \
     gpg --batch --keyserver hkp://pgp.mit.edu:80 --recv-keys "$key" ; \
   done \
+  && ( \
+    curl -fsSLO --compressed "https://unofficial-builds.nodejs.org/download/release/v$NODE_VERSION/node-v$NODE_VERSION-linux-$ARCH-musl.tar.xz" \
+    && curl -fsSLO --compressed "https://unofficial-builds.nodejs.org/download/release/v$NODE_VERSION/SHASUMS256.txt" \
```

Comment:
Downloading the shasums file at the same time as the tar doesn't really increase the authenticity of the artifact. If someone is able to compromise the `unofficial-builds` server (or the connection to it), then they can easily change both files to match.

We usually recommend downloading the sha file separately (like during `update.sh`) so that it can be embedded in the Dockerfile. See https://github.com/docker-library/official-images/tree/3b512a89a9719efa190d5938ea3ca4ca5291d68c#image-build

There is also a bug in that a failed `sha256sum` (or `tar`, `rm`, or `ln`) could leave the tar, shasum, or extracted files intact in the image while still then building from source. For `RUN` lines that need more complex logic, you might want to move to using `set -eu` with semicolons rather than `&&` (https://github.com/docker-library/mongo/issues/313#issuecomment-438018953).

Overall, these new builds will be nice to have!

Reply:
So, it should really only try building from source if neither file exists, since a sha or other failure indicates a different problem and the image should appropriately fail to build.

Reply:
I don't think these comments have been addressed yet. 😉

Reply:
True, but this approach is in line with the other image so maybe it warrants a separate discussion.

Reply:
I'm not following -- which "other image" are you referring to?

Reply:
The variants of this repo I mean.

Reply:
@LaurentGoderre, you are correct that the other images download the shasums file at the same time as the node release (source or pre-compiled), but they use a gpg key to verify the shasums before trusting it. As far as I can tell, this is not something that can be done anytime soon on these unofficial builds, so we definitely need to embed the shasum.

https://github.com/nodejs/docker-node/blob/3be8bf6c6adc14f475fb45188eef5ac74a7d112f/Dockerfile-alpine.template#L27-L30
https://github.com/nodejs/docker-node/blob/3be8bf6c6adc14f475fb45188eef5ac74a7d112f/Dockerfile-debian.template#L27-L30

Reply:
We also need to address when these lines fail:

> There is also a bug in that a failed `sha256sum` (or `tar`, `rm`, or `ln`) could leave the tar, shasum, or extracted files intact in the image while still then building from source. For `RUN` lines that need more complex logic, you might want to move to using `set -eu` with semicolons rather than `&&` (https://github.com/docker-library/mongo/issues/313#issuecomment-438018953).

If the tar downloads (and fails to shasum or something), it should consider that as the only possible route for the given version, rather than falling back to building, so that the build failure can be noticed and the unofficial artifacts fixed.

Reply:
> We also need to address when these lines fail:
> 
> > There is also a bug in that a failed `sha256sum` (or `tar`, `rm`, or `ln`) could leave the tar, shasum, or extracted files intact in the image while still then building from source. For `RUN` lines that need more complex logic, you might want to move to using `set -eu` with semicolons rather than `&&` ([docker-library/mongo#313 (comment)](https://github.com/docker-library/mongo/issues/313#issuecomment-438018953)).
> 
> If the tar downloads (and fails to shasum or something), it should consider that as the only possible route for the given version, rather than falling back to building, so that the build failure can be noticed and the unofficial artifacts fixed.

Reply:
Done!!! Only thing left is to modify the update script to embed the shasum.

Reply:
@yosifkit @tianon all done!

Reply:
The current implementation still suffers here -- as written, it will always attempt to download an unofficial-builds tarball whether one is expected to exist or not, and even if we're on an architecture where we expect to have such a tarball, if it does not exist, we'll happily throw a harmless error in the logs nobody will check followed by downloading the source and building instead.  Not a bad strategy from a "resilient builds" perspective, but it doesn't help catch issues with the unofficial-builds project.

Here's an alternative approach (created from `10/alpine/Dockerfile` to illustrate so I'd have something concrete I could test/verify):

```dockerfile
FROM alpine:3.9

ENV NODE_VERSION 10.16.1

RUN set -eux; \
    addgroup -g 1000 node; \
    adduser -u 1000 -G node -s /bin/sh -D node; \
    apk add --no-cache \
        libstdc++ \
    ; \
    apk add --no-cache --virtual .build-deps \
        curl \
    ; \
    apkArch="$(apk --print-arch)"; \
    url=; checksum=; \
    case "$apkArch" in \
        x86_64) \
            url="https://unofficial-builds.nodejs.org/download/release/v$NODE_VERSION/node-v$NODE_VERSION-linux-x64-musl.tar.xz"; \
            checksum='c5b5adc27dfe1c9bded56b30158f9a10c617c93dde1f74febc19866dc5ff487c'; \
            ;; \
    esac; \
    if [ -n "$url" ] && [ -n "$checksum" ]; then \
        curl -fsSL -o node.txz --compressed "$url"; \
        echo "$checksum *node.txz" | sha256sum -c -; \
        tar -xJf node.txz -C /usr/local --strip-components=1 --no-same-owner; \
        rm node.txz; \
    else \
        apk add --no-cache --virtual .build-deps-full \
            binutils-gold \
            g++ \
            gcc \
            gnupg \
            libgcc \
            linux-headers \
            make \
            python \
        ; \
# gpg keys listed at https://github.com/nodejs/node#release-keys
        for key in \
            94AE36675C464D64BAFA68DD7434390BDBE9B9C5 \
            FD3A5288F042B6850C66B31F09FE44734EB7990E \
            71DCFD284A79C3B38668286BC97EC7A07EDE3FC1 \
            DD8F2338BAE7501E3DD5AC78C273792F7D83545D \
            C4F0DFFF4E8C1A8236409D08E73BC641CC11F4C8 \
            B9AE9905FFD7803F25714661B63B535A4C206CA9 \
            77984A986EBC2AA786BC0F66B01FBB92821C587A \
            8FCCA13FEF1D0C2E91008E09770F7A9A5AE15600 \
            4ED778F539E3634C779C87C6D7062848A1AB005C \
            A48C2BEE680E841632CD4E44F07496B3EB3C1762 \
            B9E2F5981AA6E0CD28160D9FF13993A75599653C \
        ; do \
            gpg --batch --keyserver hkp://p80.pool.sks-keyservers.net:80 --recv-keys "$key" || \
            gpg --batch --keyserver hkp://ipv4.pool.sks-keyservers.net --recv-keys "$key" || \
            gpg --batch --keyserver hkp://pgp.mit.edu:80 --recv-keys "$key" ; \
        done; \
        curl -fsSLO --compressed "https://nodejs.org/dist/v$NODE_VERSION/node-v$NODE_VERSION.tar.xz"; \
        curl -fsSLO --compressed "https://nodejs.org/dist/v$NODE_VERSION/SHASUMS256.txt.asc"; \
        gpg --batch --decrypt --output SHASUMS256.txt SHASUMS256.txt.asc; \
        grep " node-v$NODE_VERSION.tar.xz\$" SHASUMS256.txt | sha256sum -c -; \
        tar -xf "node-v$NODE_VERSION.tar.xz"; \
        cd "node-v$NODE_VERSION"; \
        ./configure; \
        make -j$(getconf _NPROCESSORS_ONLN) V=; \
        make install; \
        apk del --no-network .build-deps-full; \
        cd ..; \
        rm -Rf "node-v$NODE_VERSION"; \
        rm "node-v$NODE_VERSION.tar.xz" SHASUMS256.txt.asc SHASUMS256.txt; \
    fi; \
    apk del --no-network .build-deps; \
# basic sanity/smoke test
    node --version

ENV YARN_VERSION 1.17.3

RUN apk add --no-cache --virtual .build-deps-yarn curl gnupg tar \
  && for key in \
    6A010C5166006599AA17F08146C2130DFD2497F5 \
  ; do \
    gpg --batch --keyserver hkp://p80.pool.sks-keyservers.net:80 --recv-keys "$key" || \
    gpg --batch --keyserver hkp://ipv4.pool.sks-keyservers.net --recv-keys "$key" || \
    gpg --batch --keyserver hkp://pgp.mit.edu:80 --recv-keys "$key" ; \
  done \
  && curl -fsSLO --compressed "https://yarnpkg.com/downloads/$YARN_VERSION/yarn-v$YARN_VERSION.tar.gz" \
  && curl -fsSLO --compressed "https://yarnpkg.com/downloads/$YARN_VERSION/yarn-v$YARN_VERSION.tar.gz.asc" \
  && gpg --batch --verify yarn-v$YARN_VERSION.tar.gz.asc yarn-v$YARN_VERSION.tar.gz \
  && mkdir -p /opt \
  && tar -xzf yarn-v$YARN_VERSION.tar.gz -C /opt/ \
  && ln -s /opt/yarn-v$YARN_VERSION/bin/yarn /usr/local/bin/yarn \
  && ln -s /opt/yarn-v$YARN_VERSION/bin/yarnpkg /usr/local/bin/yarnpkg \
  && rm yarn-v$YARN_VERSION.tar.gz.asc yarn-v$YARN_VERSION.tar.gz \
  && apk del .build-deps-yarn

COPY docker-entrypoint.sh /usr/local/bin/
ENTRYPOINT ["docker-entrypoint.sh"]

CMD [ "node" ]
```

And the `diff`, which should make the changes I've made more obvious:

```diff
$ GIT_PAGER=cat git diff --ignore-all-space
diff --git a/10/alpine/Dockerfile b/10/alpine/Dockerfile
index c165a6c..b653f4e 100644
--- a/10/alpine/Dockerfile
+++ b/10/alpine/Dockerfile
@@ -2,13 +2,31 @@ FROM alpine:3.9
 
 ENV NODE_VERSION 10.16.1
 
-RUN addgroup -g 1000 node \
-    && adduser -u 1000 -G node -s /bin/sh -D node \
-    && apk add --no-cache \
+RUN set -eux; \
+    addgroup -g 1000 node; \
+    adduser -u 1000 -G node -s /bin/sh -D node; \
+    apk add --no-cache \
         libstdc++ \
-    && apk add --no-cache --virtual .build-deps \
-        binutils-gold \
+    ; \
+    apk add --no-cache --virtual .build-deps \
         curl \
+    ; \
+    apkArch="$(apk --print-arch)"; \
+    url=; checksum=; \
+    case "$apkArch" in \
+        x86_64) \
+            url="https://unofficial-builds.nodejs.org/download/release/v$NODE_VERSION/node-v$NODE_VERSION-linux-x64-musl.tar.xz"; \
+            checksum='c5b5adc27dfe1c9bded56b30158f9a10c617c93dde1f74febc19866dc5ff487c'; \
+            ;; \
+    esac; \
+    if [ -n "$url" ] && [ -n "$checksum" ]; then \
+        curl -fsSL -o node.txz --compressed "$url"; \
+        echo "$checksum *node.txz" | sha256sum -c -; \
+        tar -xJf node.txz -C /usr/local --strip-components=1 --no-same-owner; \
+        rm node.txz; \
+    else \
+        apk add --no-cache --virtual .build-deps-full \
+            binutils-gold \
             g++ \
             gcc \
             gnupg \
@@ -16,8 +34,9 @@ RUN addgroup -g 1000 node \
             linux-headers \
             make \
             python \
+        ; \
 # gpg keys listed at https://github.com/nodejs/node#release-keys
-  && for key in \
+        for key in \
             94AE36675C464D64BAFA68DD7434390BDBE9B9C5 \
             FD3A5288F042B6850C66B31F09FE44734EB7990E \
             71DCFD284A79C3B38668286BC97EC7A07EDE3FC1 \
@@ -33,20 +52,24 @@ RUN addgroup -g 1000 node \
             gpg --batch --keyserver hkp://p80.pool.sks-keyservers.net:80 --recv-keys "$key" || \
             gpg --batch --keyserver hkp://ipv4.pool.sks-keyservers.net --recv-keys "$key" || \
             gpg --batch --keyserver hkp://pgp.mit.edu:80 --recv-keys "$key" ; \
-  done \
-    && curl -fsSLO --compressed "https://nodejs.org/dist/v$NODE_VERSION/node-v$NODE_VERSION.tar.xz" \
-    && curl -fsSLO --compressed "https://nodejs.org/dist/v$NODE_VERSION/SHASUMS256.txt.asc" \
-    && gpg --batch --decrypt --output SHASUMS256.txt SHASUMS256.txt.asc \
-    && grep " node-v$NODE_VERSION.tar.xz\$" SHASUMS256.txt | sha256sum -c - \
-    && tar -xf "node-v$NODE_VERSION.tar.xz" \
-    && cd "node-v$NODE_VERSION" \
-    && ./configure \
-    && make -j$(getconf _NPROCESSORS_ONLN) V= \
-    && make install \
-    && apk del .build-deps \
-    && cd .. \
-    && rm -Rf "node-v$NODE_VERSION" \
-    && rm "node-v$NODE_VERSION.tar.xz" SHASUMS256.txt.asc SHASUMS256.txt
+        done; \
+        curl -fsSLO --compressed "https://nodejs.org/dist/v$NODE_VERSION/node-v$NODE_VERSION.tar.xz"; \
+        curl -fsSLO --compressed "https://nodejs.org/dist/v$NODE_VERSION/SHASUMS256.txt.asc"; \
+        gpg --batch --decrypt --output SHASUMS256.txt SHASUMS256.txt.asc; \
+        grep " node-v$NODE_VERSION.tar.xz\$" SHASUMS256.txt | sha256sum -c -; \
+        tar -xf "node-v$NODE_VERSION.tar.xz"; \
+        cd "node-v$NODE_VERSION"; \
+        ./configure; \
+        make -j$(getconf _NPROCESSORS_ONLN) V=; \
+        make install; \
+        apk del --no-network .build-deps-full; \
+        cd ..; \
+        rm -Rf "node-v$NODE_VERSION"; \
+        rm "node-v$NODE_VERSION.tar.xz" SHASUMS256.txt.asc SHASUMS256.txt; \
+    fi; \
+    apk del --no-network .build-deps; \
+# basic sanity/smoke test
+    node --version
 
 ENV YARN_VERSION 1.17.3
 
```

If you like this approach, I'd be happy to put together a proper PR.